### DISCUSSION TITLE
refresh useFetch with query parameter

### DISCUSSION THREAD

**Discussion by julienhohm** [nuxt/nuxt#18453] [Questions] [answered]
Hello,

I am certainly making a mistake, but I can't find where in the code above:
```
const championshipId = ref("");
const { data: championships } = await useFetch("/api/admin/championships");

const { data: games, refresh } = await useFetch(() => "/api/admin/games", {
  query: {
    championshipId: championshipId.value,
  },
});

function getGames(id) {
  championshipId.value = id;
  refresh();
}
```

When I click on a `li`, it triggers the `getGames` function and I want to retrieve related data.
However, the debugger shows the url called is `http://localhost:3000/api/admin/games?championshipId` and not `http://localhost:3000/api/admin/games?championshipId=63bc19c69aa2921d335f3f0f` as wanted and I get no data.

I found two solutions:
```
const { data: games, refresh } = await useAsyncData("games", () =>
  $fetch("/api/admin/games", {
    query: {
      championshipId: championshipId.value,
    },
  })
);
```

or 

```
const { data: games, refresh } = await useFetch(
  () => `/api/admin/games?championshipId=${championshipId.value}`
);
```

Is there no solution to refresh useFetch with the query options or do I make something wrong?

Thanks a lot.

**Comment by TheAlexLichter** [comment]
Pass `championshipId` as `ref`, not as value. This will make `useFetch` watch it right away.


```
const championshipId = ref(1);
const { data: championships } = await useFetch("/api/admin/championships");

const { data: games, refresh } = await useAsyncData("games", () =>
  $fetch("/api/admin/games", {
    query: {
      championshipId // not .value
    },
  })
);

```

**Comment by julienhohm** [comment]
Hello,

Sorry for the delay.
It still not working as you can see the error below.
```
const {
  data: games,
  pending,
  refresh,
} = await useFetch(() => "/api/admin/games", { query: { championshipId } });
```
![image](https://user-images.githubusercontent.com/32843802/214794116-9460197d-e6b9-4fbd-9d6d-ac6684782100.png)

Maybe it's because the declaration is wrong as my championshipId is a uid: 63ce9c3c80ee52df39f38d6a

**Reply by TheAlexLichter** [reply]
Please share a reproduction then via stackblitz :)

**Comment by julienhohm** [comment]
Was coding a sample on stackblitz, and I noticed that I was working on 3.0.11 RC on my side.
After upgrade to 3.1.1, it's working fine with your solution! My bad I did not found the issue before open this discussion...

Anyway, don't know why, there is 2 requests on the click, one canceled and one with data returned:
![image](https://user-images.githubusercontent.com/32843802/214873253-bece593f-f687-44de-ae67-3944505dbb63.png)

You can find the same problem on this example:
[https://stackblitz.com/edit/nuxt-starter-rjvrva?file=app.vue,server%2Fapi%2Fusers.get.ts,server%2Fapi%2Fuser.get.ts](https://stackblitz.com/edit/nuxt-starter-rjvrva?file=app.vue,server%2Fapi%2Fusers.get.ts,server%2Fapi%2Fuser.get.ts)

**Reply by julienhohm** [reply]
Hello,

Sorry for the delay, but I realize that my last question is unsolved.
To resume: when I click on one of buttons (surrounded in red) on the left sidebar, I want to fetch data and fill the right side (surrounded in blue)
![image](https://github.com/nuxt/nuxt/assets/32843802/88251d33-aecf-428c-8660-1a0c797dd6f8)

It's working, but the Newtork tab of Google Chrome is saying 2 fetchs are done: one canceled then the second performed.
![image](https://github.com/nuxt/nuxt/assets/32843802/c3f7a0fd-580b-4149-9274-51ac9037c6ae)

Is that normal? 
Here is my code in my parent component :

```
const championshipId = ref("");

function selectChampionship(id: string) {
  championshipId.value = id;
}

<template>
    ...
   <AdminGames :championshipId="championshipId"/>
    ...
</template>
```

Here is my code in my child component.
```
const props = defineProps({
  championshipId: {
    required: true,
    type: String,
  },
});

const championshipId = ref("");

const {
  data: games,
  pending,
  refresh,
} = await useFetch<(Game & { homeClub: Club; visitorClub: Club })[] | null>(
  "/api/admin/games",
  {
    query: { championshipId },
  }
);

watch(
  () => props.championshipId,
  () => {
    championshipId.value = props.championshipId;
    refresh();
  }
);
```

Thanks for your help!

**Reply by darioferderber** [reply, marked as answer]
@julienhohm, you don't need to call `refresh()`, it will work just fine without that.

https://nuxt.com/docs/api/composables/use-fetch
> All fetch options can be given a computed or ref value. These will be watched and new requests made automatically with any new values if they are updated.

Basically, the first request is triggered automatically, and the second one is triggered with `refresh()`.

Cheers!

**Reply by julienhohm** [reply]
Hello @darioferderber 

Thanks for your answer! It helps a lot! I finally solve my problem, using the code above:

```typescript
const championshipId = computed(() => props.championshipId);

const { data: games, pending } = await useFetch<
  (Game & { homeClub: Club; visitorClub: Club })[] | null
>("/api/admin/games", { query: { championshipId } });
```

It's just a shame not to be able to use props directly, or maybe I make a mistake somewhere?

Thanks again!

**Reply by TheAlexLichter** [reply]
You should be able to use sth. like: `{ query: { championshipId: props.championshipId } }`. You can't destructure props (yet) because reactivity will be lost then.

**Reply by julienhohm** [reply]
Hello manniL,

In fact, I already tried this syntax, but it's not working: no fetch is performed when I click on buttons. However, it works with this one:
```typescript
const { data: games, pending } = await useFetch<
  (Game & { homeClub: Club; visitorClub: Club })[] | null
>(() => `/api/admin/games?championshipId=${props.championshipId}`);
```

Don't know why the query parameter (also tried params in case) does not work.

**Reply by darioferderber** [reply]
@manniL but `props.championshipId` is not a `ref`, so I think it needs to be a `computed` property.
You can't watch primitive value, type of `props.championshipId` will be `number` or `string`.

**Reply by julienhohm** [reply]
After a quick review of useFetch and defineProps documentation, there is no mention that props.something is a ref, as you said.

It seems to be all work as it should. All my issues solved 😄 

Thanks all, I close this discussion resolved.

**Reply by TheAlexLichter** [reply]
@darioferderber True, you are right 😋👍🏻

**Reply by darioferderber** [reply]
@manniL 😋
